pkg/sentry/fsimpl/fuse/init.go

```diff
@@ -0,0 +1,245 @@
+// Copyright 2020 The gVisor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fuse
+
+import (
+	"gvisor.dev/gvisor/pkg/abi/linux"
+	"gvisor.dev/gvisor/pkg/sentry/kernel/auth"
+)
+
+// consts used by FUSE_INIT negotiation.
+const (
+	// fuseMaxMaxPages is the maximum value for MaxPages received in InitOut.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMaxMaxPages = 256
+
+	// Maximum value for the time granularity for file time stamps, 1s.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMaxTimeGranNs = 1000000000
+
+	// Minimum value for MaxWrite.
+	// Follow the same behavior as unix fuse implementation.
+	fuseMinMaxWrite = 4096
+
+	// Temporary default value for max readahead, 128kb.
+	fuseDefaultMaxReadahead = 131072
+
+	// The FUSE_INNT_IN flags sent to the daemon.
+	// TODO(gvisor.dev/issue/3199): complete the flags.
+	fuseDefaultInitFlags = linux.FUSE_MAX_PAGES
+)
+
+// Adjustable maximums for Connection's cogestion control parameters.
+// Used as the upperbound of the config values.
+// Currently we do not support adjustment to them.
+var (
+	MaxUserBackgroundRequest   uint16 = fuseDefaultMaxBackground
+	MaxUserCongestionThreshold uint16 = fuseDefaultCongestionThreshold
+)
+
+// InitSend sends a FUSE_INIT request.
+func (fs *filesystem) InitSend(creds *auth.Credentials, pid uint32) error {
```

Comment:
Why is `InitSend` and `InitRecv` defined on `filesystem` and not `Connection`?

If you look at the body of these functions, all they do is modify the underlying connection through `fs.conn`. It seems they logically belong on the `Connection`.

Reply:
@mrahatm semantically it makes more sense for those to be in the `filesystem`. `InitSend` uses the connection to create a new request and to call the server. Every other operator does that as part of the filesystem operations and I don't think Init should be an exception.

Also I wanted `connection` to be as small as possible. Responsible for just communicating between the sentry and the server. All other logic of how init should build the request and when or how to call and process the response should belong in `filesystem` and not the Connection itself imo.

Reply:
I personally think it more intuitive to kindle ask the `connection` to initializes itself. Since FUSE_INIT is unique to FUSE and is not a normal file system operation, and one initialization exactly matches one `connection` instance

Reply:
I think this can be argued both ways at a high level, but practically, all FUSE_INIT does at the moment is set a bunch of fields on the connection. 

Having this as part of the filesystem violates the principle of [encapsulation](https://en.wikipedia.org/wiki/Encapsulation_(computer_programming)): that objects manage their own internal state. Why is the filesystem responsible for configuring internals of the connection?

If we really think these functions belong on the filesystem, the configurations should also be moved to the filesystem. If there's no good case for doing so, that's probably a good indication these belong to the connection.

I'm happy with them being defined on either the connection or the filesystem, as long as the configurations are moved appropriately.